### ISSUE TITLE
cnn模型构建的问题

### ISSUE THREAD

**Issue by 15950552800** [ypwhs/captcha_break#36]
在构建cnn模型的时候：
input_tensor = Input((height, width, 3))
    x = input_tensor
    for i, n_cnn in enumerate([2, 2, 2, 2, 2]):
      ############
        for j in range(n_cnn):
            这里为什么要循环两次呢？
            x = Conv2D(32 * 2 ** min(i, 3), kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
            x = BatchNormalization()(x)
            x = Activation('relu')(x)
        x = MaxPooling2D(2)(x)

    x = Flatten()(x)
    x = [Dense(n_class, activation='softmax', name='c%d' % (i + 1))(x) for i in range(n_len)]
    model = Model(inputs=input_tensor, outputs=x)

**Comment by ypwhs** [comment]
因为我想构建 [2, 2, 2, 2, 2] 的结构，这样写比较方便。

你也可以写成这样：

```py
x = Conv2D(32, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = Conv2D(32, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = MaxPooling2D(2)(x)
x = Conv2D(64, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = Conv2D(64, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = MaxPooling2D(2)(x)
x = Conv2D(128, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = Conv2D(128, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = MaxPooling2D(2)(x)
x = Conv2D(256, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = Conv2D(256, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = MaxPooling2D(2)(x)
x = Conv2D(256, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = Conv2D(256, kernel_size=3, padding='same', kernel_initializer='he_uniform')(x)
x = BatchNormalization()(x)
x = Activation('relu')(x)
x = MaxPooling2D(2)(x)
```

这样写比较清晰，但是想调成其他结构比较费劲。

**Comment by 15950552800** [comment]
感谢您的回复！我能不能请问一下您是如何确定的模型呢？比如说为什么要用相同的卷积核卷积两次呢？是会加速模型收敛吗？

**Comment by 15950552800** [comment]
感谢您的回复！我能不能请问一下您是如何确定的模型结构呢？比如说为什么要用相同的卷积核卷积两次呢？是会加速模型收敛吗？

**Comment by ypwhs** [comment]
> 感谢您的回复！我能不能请问一下您是如何确定的模型结构呢？比如说为什么要用相同的卷积核卷积两次呢？是会加速模型收敛吗？

看论文，学习别人的模型结构设计方法，针对这个问题，可以看这篇论文：

**An End-to-End Trainable Neural Network for Image-based Sequence Recognition and Its Application to Scene Text Recognition**
[https://arxiv.org/abs/1507.05717](https://arxiv.org/abs/1507.05717)
Submitted on 21 Jul 2015

如果希望设计通用的模型结构，可以看下面的[参考资料](https://github.com/ypwhs/dl-engineer-guidebook/blob/master/docs/best-models-on-imagenet.md)：

## 模型论文

参考链接：[https://paperswithcode.com/sota/image-classification-on-imagenet](https://paperswithcode.com/sota/image-classification-on-imagenet)

### VGG <a id="VGG"></a>

**Very Deep Convolutional Networks for Large-Scale Image Recognition**  
[https://arxiv.org/abs/1409.1556](https://arxiv.org/abs/1409.1556)  
Submitted on 4 Sep 2014

### ResNet <a id="ResNet"></a>

**Deep Residual Learning for Image Recognition**  
[https://arxiv.org/abs/1512.03385](https://arxiv.org/abs/1512.03385)  
Submitted on 10 Dec 2015

### InceptionV3 <a id="InceptionV3"></a>

**Rethinking the Inception Architecture for Computer Vision**  
[https://arxiv.org/abs/1512.00567](https://arxiv.org/abs/1512.00567)  
Submitted on 2 Dec 2015

### InceptionResNetV2 <a id="InceptionResNetV2"></a>

**Inception-v4, Inception-ResNet and the Impact of Residual Connections on Learning**  
[https://arxiv.org/abs/1602.07261](https://arxiv.org/abs/1602.07261)  
Submitted on 23 Feb 2016

### DenseNet <a id="DenseNet"></a>

**Densely Connected Convolutional Networks**  
[https://arxiv.org/abs/1608.06993](https://arxiv.org/abs/1608.06993)  
Submitted on 25 Aug 2016

### Xception <a id="Xception"></a>

**Xception: Deep Learning with Depthwise Separable Convolutions**  
[https://arxiv.org/abs/1610.02357](https://arxiv.org/abs/1610.02357)  
Submitted on 7 Oct 2016

### NASNet <a id="NASNet"></a>

**Neural Architecture Search with Reinforcement Learning**  
[https://arxiv.org/abs/1611.01578](https://arxiv.org/abs/1611.01578)  
Submitted on 5 Nov 2016

### AmoebaNet

**Regularized Evolution for Image Classifier Architecture Search**  
[https://arxiv.org/abs/1802.01548](https://arxiv.org/abs/1802.01548)  
Submitted on 5 Feb 2018

### EfficientNet

**EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks**  
[https://arxiv.org/abs/1905.11946](https://arxiv.org/abs/1905.11946)  
Submitted on 28 May 2019

**Comment by 15950552800** [comment]
感谢您的回复，对我帮助很大，谢谢！
